### DISCUSSION TITLE
get_car_data() 2021-Yeda-Race-Verstappen-37

### DISCUSSION THREAD

**Discussion by formulatimer** [theOehrly/Fast-F1#519] [Q&A] [answered]
I am trying to plot Verstappen's lap 37 telemetry from the 2021 Yeda Race. (Yes, it is the lap where HAM and VER collide and the suspicious brake test of VER occurs.)

I am able to plot Verstappen's lap 35 telemetry, but when I change the lap to 37, I receive an error
```
lap_telemetry_ = lap_selected.get_car_data().add_distance()

line 3413, in get_car_data
car_data = self.session.car_data[self['DriverNumber']] 
KeyError: nan
```

**Does it mean that the telemetry for that lap is not available?**
In that case, **how do I know if the telemetry for a lap is available before trying to plot it?**

Here is the code:
```
session = ff1.get_session(2021, 'Yeda', 'Race')
session.load(laps=True, telemetry=True, weather=False, messages=False, livedata=None)
#lap_selected=session.laps.pick_driver('VER').pick_laps(35).pick_fastest()
lap_selected=session.laps.pick_driver('VER').pick_laps(37).pick_fastest()
lap_telemetry = lap_selected.get_car_data().add_distance()
```

**Comment by theOehrly** [comment, marked as answer]
Your issue is that you are using ``.pick_fastest()`` after you pick the lap number. By default, ``.pick_fastest()`` will only return laps that are marked as personal best laps (i.e. they were faster than any previous lap and not deleted). Because lap 37 is not a personal best lap for VER, no lap is returned here. Therefore, you are then trying to get the telemetry of no lap which of course returns no data.
In short, remove the additional ``.pick_fastest()`` call and everything will work.

**Reply by formulatimer** [reply]
Hi,

I a using `pick_fastest()` because it returns `<class 'fastf1.core.Lap'>`. If I remove `pick_fastest()`, it returns `<class 'fastf1.core.Laps'>`.
When I ty to obtain the LapTime in seconds
`session.laps.pick_driver('VER').pick_laps(37).pick_fastest()['LapTime'].total_seconds()` it returns a float but if I remove `pick_fastest()` it fails because `'Series' object has no attribute 'total_seconds'`.
**Is there any function that converts from `<class 'fastf1.core.Laps'>` to `<class 'fastf1.core.Lap'>`?**

**Reply by theOehrly** [reply]
FastF1's ``Lap`` and ``Laps`` classes (and others) are subclasses of pandas' Series and DataFrames. Specifically, ``Laps`` is basically a ``DataFrame`` with additional functionality. Therefore you can use all the Pandas functionality as well. 
To get a ``Lap`` from a ``Laps`` object, use for example ``.iloc[0]`` instead of using ``.pick_fastest()`` as you are currently doing.
Relevant links to the pandas documentation:
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html
https://pandas.pydata.org/docs/reference/frame.html

**Reply by formulatimer** [reply]
Thanks,

`.iloc[0]` works for me. Other option to convert from Laps to Lap is using `.iterlaps()`

However, I think that `.pick_laps()` should return a `Lap` if only one lap is selected. Other option is to create a `.pick_lap()` function but i believe that the first option is easy to implement adding an `if` and `.iloc[0]`

Do you think it is worthy?

**Reply by theOehrly** [reply]
No, this would not be a good solution. I will give you an example why.
Assume that you want to get all outlaps of all drivers who exited the box on lap 10:
```
session.laps.pick_box_laps(which='out').pick_laps(10)
```
Now, depending on the lap number only, you will either get no laps, one lap or multiple laps. But generally, you don't know how many drivers pitted on a given lap. So now you have no idea whether you'll get a ``Lap`` object or a ``Laps`` object. There are various cases where it is not generally possible to know in advance how many, if any, laps will be returned. A DataFrame can be empty or contain one or multiple entries. Therefore, it covers all possible cases and even more important, the return type is consistent.
